src/compositor/hwcimage.cpp

```diff
@@ -0,0 +1,523 @@
+        xnode->setMatrix(reverseTransform());
+        qsgnode_set_description(xnode, QStringLiteral("hwc-reverse-xform"));
+        xnode->appendChildNode(contentNode);
+        hwcNode->appendChildNode(xnode);
+        hwcNode->update(contentNode, contentNode->handle());
+        return hwcNode;
+    }
+
+    return 0;
+}
+
+
+// from hybris_nativebuffer.h in libhybris
+#define HYBRIS_USAGE_SW_WRITE_RARELY    0x00000020
+#define HYBRIS_USAGE_HW_TEXTURE         0x00000100
+#define HYBRIS_USAGE_HW_COMPOSER        0x00000800
+#define HYBRIS_PIXEL_FORMAT_RGBA_8888   1
+#define HYBRIS_PIXEL_FORMAT_RGB_888     3
+#define HYBRIS_PIXEL_FORMAT_BGRA_8888   5
+
+#define EGL_NATIVE_BUFFER_HYBRIS        0x3140
+
+extern "C" {
+    typedef EGLBoolean (EGLAPIENTRYP _eglHybrisCreateNativeBuffer)(EGLint width, EGLint height, EGLint usage, EGLint format, EGLint *stride, EGLClientBuffer *buffer);
+    typedef EGLBoolean (EGLAPIENTRYP _eglHybrisLockNativeBuffer)(EGLClientBuffer buffer, EGLint usage, EGLint l, EGLint t, EGLint w, EGLint h, void **vaddr);
+    typedef EGLBoolean (EGLAPIENTRYP _eglHybrisUnlockNativeBuffer)(EGLClientBuffer buffer);
+    typedef EGLBoolean (EGLAPIENTRYP _eglHybrisReleaseNativeBuffer)(EGLClientBuffer buffer);
+
+    typedef void (EGLAPIENTRYP _glEGLImageTargetTexture2DOES)(GLenum target, EGLImageKHR image);
+    typedef EGLImageKHR (EGLAPIENTRYP _eglCreateImageKHR)(EGLDisplay dpy, EGLContext ctx, EGLenum target, EGLClientBuffer buffer, const EGLint *attribs);
+    typedef EGLBoolean (EGLAPIENTRYP _eglDestroyImageKHR)(EGLDisplay dpy, EGLImageKHR image);
+}
+
+_glEGLImageTargetTexture2DOES glEGLImageTargetTexture2DOES = 0;
+_eglCreateImageKHR eglCreateImageKHR = 0;
+_eglDestroyImageKHR eglDestroyImageKHR = 0;
+
+_eglHybrisCreateNativeBuffer eglHybrisCreateNativeBuffer = 0;
+_eglHybrisLockNativeBuffer eglHybrisLockNativeBuffer = 0;
+_eglHybrisUnlockNativeBuffer eglHybrisUnlockNativeBuffer = 0;
+_eglHybrisReleaseNativeBuffer eglHybrisReleaseNativeBuffer = 0;
+
+static void hwcimage_initialize()
+{
+    static bool initialized = false;
+    if (initialized)
+        return;
+    initialized = true;
+
+    glEGLImageTargetTexture2DOES = (_glEGLImageTargetTexture2DOES) eglGetProcAddress("glEGLImageTargetTexture2DOES");
+    eglCreateImageKHR = (_eglCreateImageKHR) eglGetProcAddress("eglCreateImageKHR");
+    eglDestroyImageKHR = (_eglDestroyImageKHR) eglGetProcAddress("eglDestroyImageKHR");
+    eglHybrisCreateNativeBuffer = (_eglHybrisCreateNativeBuffer) eglGetProcAddress("eglHybrisCreateNativeBuffer");
+    eglHybrisLockNativeBuffer = (_eglHybrisLockNativeBuffer) eglGetProcAddress("eglHybrisLockNativeBuffer");
+    eglHybrisUnlockNativeBuffer = (_eglHybrisUnlockNativeBuffer) eglGetProcAddress("eglHybrisUnlockNativeBuffer");
+    eglHybrisReleaseNativeBuffer = (_eglHybrisReleaseNativeBuffer) eglGetProcAddress("eglHybrisReleaseNativeBuffer");
+}
+
+QSGTexture *HwcImageTexture::create(const QImage &image)
```

Comment:
On phone I've had this instantiate a seemingly blank texture.  It may have been size related as it did work for some application backgrounds but not home backgrounds.